File: daily-temperatures.py
class Solution:
    # stack: O(n), O(n)
    def dailyTemperatures(self, T: List[int]) -> List[int]:
        stack = []
        warm = [0 for _ in range(len(T))]
        
        for idx, temp in enumerate(T):
            if not stack or stack[-1][1] >= temp:
                stack.append((idx, temp))                
            elif stack[-1][1] < temp:
                while stack and stack[-1][1] < temp:
                    top = stack.pop()
                    warm[top[0]] = idx - top[0]
                stack.append((idx, temp))
        
        return warm

File: daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Deque<Pair> stack = new ArrayDeque<>(temperatures.length);
        int[] answer = new int[temperatures.length];

        for(int i=0; i<temperatures.length; i++) {
            Pair pair = new Pair(i, temperatures[i]);
            while(stack.size() > 0 && stack.peek().temperature < temperatures[i]) {
                Pair p = stack.pop();
                answer[p.index] = i - p.index;
            }
            stack.push(pair);
        }

        return answer;
    }
}

class Pair {
    int index;
    int temperature;

    Pair(int index, int temperature) {
        this.index = index;
        this.temperature = temperature;
    }
}